src/services/patternMatcher.ts

```diff
@@ -0,0 +1,810 @@
+    // once you no longer need it.
+    export interface PatternMatcher {
+        // Used to match a candidate against the last segment of a possibly dotted pattern.  This
+        // is useful as a quick check to prevent having to compute a container before calling 
+        // "getMatches".
+        //
+        // For example, if the search pattern is "ts.c.SK" and the candidate is "SyntaxKind", then
+        // this will return a successful match, having only tested "SK" against "SyntaxKind".  At
+        // that point a call can be made to 'getMatches("SyntaxKind", "ts.compiler")', with the
+        // work to create 'ts.compiler' only being done once the first match succeeded.
+        getMatchesForLastSegmentOfPattern(candidate: string): PatternMatch[];
+
+        // Fully checks a candidate, with an optional dotted container against the search pattern.
+        // The candidate must match the last part of the search pattern, and the dotted container
+        // must match the preceding segments of the pattern.
+        getMatches(candidate: string, dottedContainer: string): PatternMatch[];
+        isDottedPattern: boolean;
+    }
+
+    // First we break up the pattern given by dots.  Each portion of the pattern between the
+    // dots is a 'Segment'.  The 'Segment' contains information about the entire section of 
+    // text between the dots, as well as information about any individual 'Words' that we 
+    // can break the segment into.  A 'Word' is simply a contiguous sequence of characters
+    // that can appear in a typescript identifier.  So "GetKeyword" would be one word, while
+    // "Get Keyword" would be two words.  Once we have the individual 'words', we break those
+    // into constituent 'character spans' of interest.  For example, while 'UIElement' is one
+    // word, it make character spans corresponding to "U", "I" and "Element".  These spans
+    // are then used when doing camel cased matches against candidate patterns.
+    interface Segment {
+        // Information about the entire piece of text between the dots.  For example, if the 
+        // text between the dots is 'GetKeyword', then TotalTextChunk.Text will be 'GetKeyword' and 
+        // TotalTextChunk.CharacterSpans will correspond to 'Get', 'Keyword'.
+        totalTextChunk: TextChunk;
+
+        // Information about the subwords compromising the total word.  For example, if the 
+        // text between the dots is 'GetFoo KeywordBar', then the subwords will be 'GetFoo' 
+        // and 'KeywordBar'.  Those individual words will have CharacterSpans of ('Get' and 
+        // 'Foo') and('Keyword' and 'Bar') respectively. 
+        subWordTextChunks: TextChunk[];
+    }
+
+    // Information about a chunk of text from the pattern.  The chunk is a piece of text, with 
+    // cached information about the character spans within in.  Character spans are used for 
+    // camel case matching.
+    interface TextChunk {

```

Comment:
So a TextChunk can represent either a Segment or a Word? Why is it not a strict hierarchy, with Segments containing Words and Words containing CharacterSpans? TextChunk seems to blur the hierarchy.